<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="sigurd">
<title>Analysis of Blastic Plasmacytoid Dendritic Cell Neoplasm (BPDCN) â€¢ sigurd</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of Blastic Plasmacytoid Dendritic Cell Neoplasm (BPDCN)">
<meta property="og:description" content="sigurd">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sigurd</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.14</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/BoneMarrow_SIGURD.html">Analysis of Blastic Plasmacytoid Dendritic Cell Neoplasm (BPDCN)</a>
    <a class="dropdown-item" href="../articles/MPN_SIGURD.html">Analysis of MPN Data</a>
    <a class="dropdown-item" href="../articles/SW_CellMixture_SIGURD.html">Cell Mixture Analysis using Seq-Well Sequencing</a>
    <a class="dropdown-item" href="../articles/TenX_CellMixture_SIGURD.html">Cell Mixture Experiment using 10X Genomics Sequencing</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">

<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Analysis of Blastic Plasmacytoid Dendritic Cell Neoplasm (BPDCN)</h1>
            
      
      
      <div class="d-none name"><code>BoneMarrow_SIGURD.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="analysis-of-blastic-plasmacytoid-dendritic-cell-neoplasm-bpdcn">Analysis of Blastic Plasmacytoid Dendritic Cell Neoplasm (BPDCN)<a class="anchor" aria-label="anchor" href="#analysis-of-blastic-plasmacytoid-dendritic-cell-neoplasm-bpdcn"></a>
</h2>
<p>Here, we compare the analysis of cells from a patient with BPDCN.</p>
<p>The original script is available here: <a href="https://github.com/petervangalen/MAESTER-2021/blob/main/4_CH_sample/4.2_Variant_Selection.R" class="external-link uri">https://github.com/petervangalen/MAESTER-2021/blob/main/4_CH_sample/4.2_Variant_Selection.R</a></p>
<p>In this setup, the cells are all from a single donor and do not have to be separated into categories.</p>
<p>Then small clones that might characterize the clonal lineages in the sample are selected.</p>
</div>
<div class="section level2">
<h2 id="loading-necessary-packages-">Loading necessary packages.<a class="anchor" aria-label="anchor" href="#loading-necessary-packages-"></a>
</h2>
<p>We load the data from MAESTER.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Libraries for SIGURD."</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Libraries for SIGURD."</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://costalab.github.io/sigurd/">sigurd</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-the-data-using-sigurd-">Loading the data using SIGURD.<a class="anchor" aria-label="anchor" href="#loading-the-data-using-sigurd-"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The design matrix contains information for the genotyping data.</span></span>
<span><span class="va">genotyping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadingMAEGATK_typewise.html">LoadingMAEGATK_typewise</a></span><span class="op">(</span>patient <span class="op">=</span> <span class="st">"BPDCN712"</span>, samples_file <span class="op">=</span> <span class="st">"/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/data/MAESTER_Reproduction.csv"</span>, type_use <span class="op">=</span> <span class="st">"Amplicon_MT"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Loading the scRNA-seq data.</span></span>
<span><span class="va">scrna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/data/BPDCN712_Seurat_with_TCR_Renamed.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="generating-a-block-list">Generating a block list<a class="anchor" aria-label="anchor" href="#generating-a-block-list"></a>
</h2>
<p>Variants that are also detected in the cell mixture data is treated as possible false positives. They are used as a blacklist and are removed from the results.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Loading the TenX Cell Mixture Genotyping.</span></span>
<span></span>
<span><span class="va">genotyping_tenx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_object.html">load_object</a></span><span class="op">(</span><span class="st">"/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/data/TenX_CellMixture_Genotyping.rds.lz4"</span><span class="op">)</span></span>
<span><span class="va">blocklist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AllelFrequencyFoldChange.html">AllelFrequencyFoldChange</a></span><span class="op">(</span><span class="va">genotyping_tenx</span>, group_of_interest <span class="op">=</span> <span class="st">"CellType"</span>, group1 <span class="op">=</span> <span class="st">"K562"</span>, group2 <span class="op">=</span> <span class="st">"BT142"</span>, maximum_foldchange <span class="op">=</span> <span class="fl">5</span>, minimum_coverage <span class="op">=</span> <span class="fl">5</span>, minimum_allele_freq <span class="op">=</span> <span class="fl">0.001</span>, maximum_allele_freq <span class="op">=</span> <span class="fl">0.999</span><span class="op">)</span><span class="op">$</span><span class="va">Variant</span></span>
<span></span>
<span><span class="co"># We add the variant chrM_1583_A_G by hand. It is identified as misleading based on downstream analysis.</span></span>
<span><span class="va">blocklist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">blocklist</span>, <span class="st">"chrM_1583_A_G"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="selecting-the-variants-of-interest">Selecting the Variants of Interest<a class="anchor" aria-label="anchor" href="#selecting-the-variants-of-interest"></a>
</h2>
<p>Now, we select the variants of interest from the loaded data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">voi.ch.sigurd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VariantSelection_TopCells.html">VariantSelection_TopCells</a></span><span class="op">(</span><span class="va">genotyping</span>, min_coverage <span class="op">=</span> <span class="fl">5</span>, quantiles <span class="op">=</span> <span class="fl">0.9</span>, thresholds <span class="op">=</span> <span class="fl">0</span>, top_cells <span class="op">=</span> <span class="fl">10</span>, top_VAF <span class="op">=</span> <span class="fl">0.5</span>, min_quality <span class="op">=</span> <span class="fl">30</span>, remove_nocall <span class="op">=</span> <span class="cn">FALSE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">voi.ch.sigurd</span> <span class="op">&lt;-</span> <span class="va">voi.ch.sigurd</span><span class="op">[</span><span class="op">!</span><span class="va">voi.ch.sigurd</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">blocklist</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">voi.ch.sigurd</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "chrM_683_G_A"   "chrM_1323_G_A"  "chrM_1415_G_A"  "chrM_2593_G_A" </span></span>
<span><span class="co">##  [5] "chrM_6205_G_A"  "chrM_6243_G_A"  "chrM_8697_G_A"  "chrM_9753_G_A" </span></span>
<span><span class="co">##  [9] "chrM_10158_T_A" "chrM_15140_G_A" "chrM_15812_G_A" "chrM_779_T_C"  </span></span>
<span><span class="co">## [13] "chrM_3988_T_C"  "chrM_6185_T_C"  "chrM_6293_T_C"  "chrM_9164_T_C" </span></span>
<span><span class="co">## [17] "chrM_9321_T_C"  "chrM_11864_T_C" "chrM_15299_T_C" "chrM_1171_A_G" </span></span>
<span><span class="co">## [21] "chrM_1222_A_G"  "chrM_2241_A_G"  "chrM_2623_A_G"  "chrM_3628_A_G" </span></span>
<span><span class="co">## [25] "chrM_4104_A_G"  "chrM_6218_A_G"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="visualisation-using-sigurd">Visualisation using SIGURD<a class="anchor" aria-label="anchor" href="#visualisation-using-sigurd"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genotyping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Filtering.html">Filtering</a></span><span class="op">(</span>SE <span class="op">=</span> <span class="va">genotyping</span>, cells_include <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scrna</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "We remove all cells not in the allow list."</span></span>
<span><span class="co">## [1] "We remove all the variants that are always NoCall."</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">genotyping</span><span class="op">)</span><span class="op">$</span><span class="va">CellType</span> <span class="op">&lt;-</span> <span class="va">scrna</span><span class="op">$</span><span class="va">CellType</span></span>
<span><span class="fu"><a href="../reference/HeatmapVoi.html">HeatmapVoi</a></span><span class="op">(</span>SE <span class="op">=</span> <span class="va">genotyping</span>, voi <span class="op">=</span> <span class="va">voi.ch.sigurd</span>, annotation_trait <span class="op">=</span> <span class="st">"CellType"</span>, sort_cells <span class="op">=</span> <span class="cn">TRUE</span>, remove_empty_cells <span class="op">=</span> <span class="cn">TRUE</span>, minimum_allele_freq <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p><img src="BoneMarrow_SIGURD_files/figure-html/Vis_SIGURD-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="cell-type-enrichment-2593ga">Cell Type Enrichment 2593G&gt;A<a class="anchor" aria-label="anchor" href="#cell-type-enrichment-2593ga"></a>
</h2>
<p>We check the cell type enrichment for the variant 2593G&gt;A.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result_enrichment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Enrichment_FisherTest.html">Enrichment_FisherTest</a></span><span class="op">(</span>se <span class="op">=</span> <span class="va">genotyping</span>, variant <span class="op">=</span> <span class="st">"chrM_2593_G_A"</span>, use_nocall <span class="op">=</span> <span class="cn">FALSE</span>, trait <span class="op">=</span> <span class="st">"CellType"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "We get a count matrix of the chrM_2593_G_A per CellType."</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">result_enrichment</span><span class="op">$</span><span class="va">Heatmap_Values</span><span class="op">)</span>, cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>, col<span class="op">=</span> <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">result_enrichment</span><span class="op">$</span><span class="va">Heatmap_Values</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="BoneMarrow_SIGURD_files/figure-html/unnamed-chunk-3-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="cell-type-enrichment-6243ga">Cell Type Enrichment 6243G&gt;A<a class="anchor" aria-label="anchor" href="#cell-type-enrichment-6243ga"></a>
</h2>
<p>We check the cell type enrichment for the variant 6243G&gt;A.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result_enrichment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Enrichment_FisherTest.html">Enrichment_FisherTest</a></span><span class="op">(</span>se <span class="op">=</span> <span class="va">genotyping</span>, variant <span class="op">=</span> <span class="st">"chrM_6243_G_A"</span>, use_nocall <span class="op">=</span> <span class="cn">FALSE</span>, trait <span class="op">=</span> <span class="st">"CellType"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "We get a count matrix of the chrM_6243_G_A per CellType."</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">result_enrichment</span><span class="op">$</span><span class="va">Heatmap_Values</span><span class="op">)</span>, cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>, col<span class="op">=</span> <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">result_enrichment</span><span class="op">$</span><span class="va">Heatmap_Values</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="BoneMarrow_SIGURD_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="enrichment-with-somatic-variants">Enrichment with somatic variants<a class="anchor" aria-label="anchor" href="#enrichment-with-somatic-variants"></a>
</h2>
<p>We get the number of 2593_G&gt;A cells that are also mutated for a different somatic variant.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">somatic_variants</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ASXL1.G642fs.1"</span>, <span class="st">"ASXL1.G642fs.2"</span>, <span class="st">"TET2.S792X"</span>, <span class="st">"TET2.Q1034X"</span>, <span class="st">"TET2.R1216X"</span>, <span class="st">"TET2.H1380Y"</span><span class="op">)</span></span>
<span><span class="va">variants</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadingRawMatrix_typewise.html">LoadingRawMatrix_typewise</a></span><span class="op">(</span>samples_file <span class="op">=</span> <span class="st">"/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/data/MAESTER_Reproduction.csv"</span>, patient <span class="op">=</span> <span class="st">"BPDCN712"</span>, variant_use <span class="op">=</span> <span class="va">somatic_variants</span>, matrix_column_separator <span class="op">=</span> <span class="st">"\t"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">genotyping</span> <span class="op">&lt;-</span> <span class="va">genotyping</span><span class="op">[</span><span class="va">voi.ch.sigurd</span>, <span class="op">]</span></span>
<span><span class="va">genotyping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Filtering.html">Filtering</a></span><span class="op">(</span><span class="va">genotyping</span>, cells_include <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">variants</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "We remove all cells not in the allow list."</span></span>
<span><span class="co">## [1] "We remove all the variants that are always NoCall."</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Filtering.html">Filtering</a></span><span class="op">(</span><span class="va">variants</span>, cells_include <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scrna</span><span class="op">)</span>, fraction_threshold <span class="op">=</span> <span class="fl">0.01</span>, reject_value <span class="op">=</span> <span class="st">"Reference"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "We remove all cells not in the allow list."</span></span>
<span><span class="co">## [1] "We assume that cells with a fraction smaller than our threshold are actually Reference."</span></span>
<span><span class="co">## [1] "We set consensus values to 1 (Reference) and fraction values to 0."</span></span>
<span><span class="co">## [1] "We do not set fractions between 0.01 and 1 to 1."</span></span>
<span><span class="co">## [1] "This way, we retain the heterozygous information."</span></span>
<span><span class="co">## [1] "We remove all the variants that are always NoCall."</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genotyping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CombineSEobjects.html">CombineSEobjects</a></span><span class="op">(</span>se_1 <span class="op">=</span> <span class="va">genotyping</span>, se_2 <span class="op">=</span> <span class="va">variants</span><span class="op">)</span></span>
<span><span class="va">genotyping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ClonalDefinition.html">ClonalDefinition</a></span><span class="op">(</span>se <span class="op">=</span> <span class="va">genotyping</span>, variants_ls <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"chrM_2593_G_A"</span><span class="op">)</span>, grouping <span class="op">=</span> <span class="cn">NULL</span>, identities <span class="op">=</span> <span class="cn">NULL</span>, explicit <span class="op">=</span> <span class="cn">TRUE</span>, explicit_not <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"chrM_6243_G_A"</span><span class="op">)</span>, explicit_min_vaf <span class="op">=</span> <span class="fl">0.01</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Explicit clone definition."</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genotyping</span> <span class="op">&lt;-</span> <span class="va">genotyping</span><span class="op">[</span>, <span class="va">genotyping</span><span class="op">$</span><span class="va">Clones</span> <span class="op">==</span> <span class="st">"C1"</span><span class="op">]</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VariantWiseFisherTest.html">VariantWiseFisherTest</a></span><span class="op">(</span>variants_list <span class="op">=</span> <span class="fu"><a href="../reference/RowWiseSplit.html">RowWiseSplit</a></span><span class="op">(</span>se <span class="op">=</span> <span class="va">genotyping</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Testing Variant: ASXL1.G642fs.1, 1 out of 6"</span></span>
<span><span class="co">## [1] "Testing Variant: ASXL1.G642fs.2, 2 out of 6"</span></span>
<span><span class="co">## [1] "Testing Variant: TET2.S792X, 3 out of 6"</span></span>
<span><span class="co">## [1] "Testing Variant: TET2.Q1034X, 4 out of 6"</span></span>
<span><span class="co">## [1] "Testing Variant: TET2.R1216X, 5 out of 6"</span></span>
<span><span class="co">## [1] "Testing Variant: TET2.H1380Y, 6 out of 6"</span></span>
<span><span class="co">## [1] "We remove the NA P values."</span></span>
<span><span class="co">## [1] "Adjusting P values using fdr."</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">res</span>, format<span class="op">=</span><span class="st">"html"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
Variant1
</th>
<th style="text-align:left;">
Variant2
</th>
<th style="text-align:right;">
P
</th>
<th style="text-align:right;">
OddsRatio
</th>
<th style="text-align:right;">
Cells_Alt_1_2
</th>
<th style="text-align:right;">
Cells_Alt_1_Ref_2
</th>
<th style="text-align:right;">
Cells_Alt_2_Ref_1
</th>
<th style="text-align:right;">
Cells_Ref_1_2
</th>
<th style="text-align:right;">
P_adj
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
ASXL1.G642fs.2
</td>
<td style="text-align:left;">
chrM_2593_G_A
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
TET2.S792X
</td>
<td style="text-align:left;">
chrM_2593_G_A
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
9
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
TET2.S792X
</td>
<td style="text-align:left;">
chrM_15299_T_C
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
TET2.Q1034X
</td>
<td style="text-align:left;">
chrM_2593_G_A
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
</tr>
</tbody>
</table>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Grasshoff, Ivan Costa.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
