<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="sigurd">
<title>Analysis of MPN Data â€¢ sigurd</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of MPN Data">
<meta property="og:description" content="sigurd">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sigurd</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.14</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/BoneMarrow_SIGURD.html">Analysis of Blastic Plasmacytoid Dendritic Cell Neoplasm (BPDCN)</a>
    <a class="dropdown-item" href="../articles/MPN_SIGURD.html">Analysis of MPN Data</a>
    <a class="dropdown-item" href="../articles/SW_CellMixture_SIGURD.html">Cell Mixture Analysis using Seq-Well Sequencing</a>
    <a class="dropdown-item" href="../articles/TenX_CellMixture_SIGURD.html">Cell Mixture Experiment using 10X Genomics Sequencing</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">

<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Analysis of MPN Data</h1>
            
      
      
      <div class="d-none name"><code>MPN_SIGURD.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="loading-necessary-packages-">Loading necessary packages.<a class="anchor" aria-label="anchor" href="#loading-necessary-packages-"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://costalab.github.io/sigurd/">sigurd</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize" class="external-link">circlize</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinblighe/EnhancedVolcano" class="external-link">EnhancedVolcano</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="analysis-of-mpn-data">Analysis of MPN data<a class="anchor" aria-label="anchor" href="#analysis-of-mpn-data"></a>
</h2>
<p>Here, we illustrate how to analysis genotyping data from MPN samples. The JAK2V617F mutation was genotyped for 6 MPN samples and 3 healthy controls.</p>
<p>This setup makes this analysis different from the previous analyses. Previously, only one sample was analysed at a time. Now, we load and analyse the samples simultaneously.</p>
</div>
<div class="section level2">
<h2 id="loading-mpn-data">Loading MPN data<a class="anchor" aria-label="anchor" href="#loading-mpn-data"></a>
</h2>
<p>Now, we load the JAK2V617F data and plot the mutated cells on the UMAP.</p>
<p>Since we are loading multiple samples, this code is more complicated than the code for the MAESTER SW sample.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Loading the scRNA-seq data.</span></span>
<span><span class="va">scrna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_object.html">load_object</a></span><span class="op">(</span><span class="st">"/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/Paper_Figures/ZenodoFiles/Zenodo_Seurat_Object.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Loading the genotyping data.</span></span>
<span><span class="co"># genotyping  &lt;- LoadingVarTrix_typewise(patient = "MPN", samples_file = "/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/data/DesignMatrix_VarTrix.csv", min_reads = 0, vcf_path = "/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/inst/extdata/JAK2V617F.vcf", type_use = "scRNAseq_Somatic", min_cells = 0, verbose = FALSE)</span></span>
<span><span class="co"># amplicon &lt;- LoadingVarTrix_typewise(patient = "MPN", samples_file = "/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/data/DesignMatrix_Amplicon.csv", min_reads = 0, vcf_path = "/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/inst/extdata/UCSC_JAK2V617F.vcf", type_use = "scRNAseq_Amplicon", min_cells = 0, verbose = FALSE)</span></span>
<span><span class="co"># genotyping  &lt;- AmpliconSupplementing(scRNAseq = genotyping, amplicon = amplicon, verbose = FALSE)</span></span>
<span></span>
<span><span class="co"># Filtering the genotyping object to remove false positives.</span></span>
<span><span class="co"># genotyping &lt;- Filtering(genotyping, min_cells_per_variant = 0, fraction_threshold = 0.21, cells_include = colnames(scrna), min_variants_per_cell = 1, reject_value = "NoCall", verbose = FALSE)</span></span>
<span></span>
<span><span class="co"># We load the provided genotyping object.</span></span>
<span><span class="va">genotyping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_object.html">load_object</a></span><span class="op">(</span><span class="st">"/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/data/MPN_JAK2V617F_Genotyping.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="combining-seurat-and-sigurd">Combining Seurat and SIGURD<a class="anchor" aria-label="anchor" href="#combining-seurat-and-sigurd"></a>
</h2>
<p>Now, we add the genotyping information to the Seurat object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scrna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SetVariantInfo.html">SetVariantInfo</a></span><span class="op">(</span>SE <span class="op">=</span> <span class="va">genotyping</span>, seurat_object <span class="op">=</span> <span class="va">scrna</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "You did not supply a vector of variants. All variants will be used."</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="visualisation">Visualisation<a class="anchor" aria-label="anchor" href="#visualisation"></a>
</h2>
<p>We visualize the genotyping information on the UMAP.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plotting the mutated cells on the UMAP.</span></span>
<span><span class="va">scrna</span><span class="op">$</span><span class="va">JAK2_p.V617F_c.1849G.T_consensus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">scrna</span><span class="op">$</span><span class="va">JAK2_p.V617F_c.1849G.T_consensus</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Ref"</span>, <span class="st">"Alt"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">jak2_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">scrna</span><span class="op">$</span><span class="va">JAK2_p.V617F_c.1849G.T_consensus</span> <span class="op">==</span> <span class="st">"Alt"</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ref_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">scrna</span><span class="op">$</span><span class="va">JAK2_p.V617F_c.1849G.T_consensus</span> <span class="op">==</span> <span class="st">"Ref"</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">all_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">scrna</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">scrna</span>, group.by <span class="op">=</span> <span class="st">"JAK2_p.V617F_c.1849G.T_consensus"</span>, reduction <span class="op">=</span> <span class="st">"INTE_UMAP"</span>, order <span class="op">=</span> <span class="cn">TRUE</span>, pt.size <span class="op">=</span> <span class="fl">1</span>, raster <span class="op">=</span> <span class="cn">FALSE</span>, na.value <span class="op">=</span> <span class="st">"grey95"</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Alt <span class="op">=</span> <span class="st">"#aa0051"</span>, Ref <span class="op">=</span> <span class="st">"#79b938"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Alt: "</span>, <span class="va">jak2_cells</span>, <span class="st">", Ref: "</span>, <span class="va">ref_cells</span>, <span class="st">", All: "</span>, <span class="va">all_cells</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"UMAP 1"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"UMAP 2"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Alt"</span>, <span class="st">"Ref"</span><span class="op">)</span>, values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#aa0051"</span>, <span class="st">"#79b938"</span><span class="op">)</span>, na.value <span class="op">=</span> <span class="st">"grey95"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Scale for <span style="color: #00BB00;">colour</span> is already present.</span></span>
<span><span class="co">## Adding another scale for <span style="color: #00BB00;">colour</span>, which will replace the existing scale.</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<p><img src="MPN_SIGURD_files/figure-html/PlottingJAK2V617F-1.png" width="384"></p>
</div>
<div class="section level2">
<h2 id="we-determine-degs-between-the-jak2v617f-and-wt-cells-">We determine DEGs between the JAK2V617F and WT cells.<a class="anchor" aria-label="anchor" href="#we-determine-degs-between-the-jak2v617f-and-wt-cells-"></a>
</h2>
<p>Now, we determine the DEGs between the JAK2V617F cells and the wild type cells for the Orthochromatic Erythroblasts.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scrna_orthoe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">scrna</span>, <span class="va">celltype</span> <span class="op">==</span> <span class="st">"Orthochromatic Erythroblasts"</span> <span class="op">&amp;</span> <span class="va">condition</span> <span class="op">!=</span> <span class="st">"HC"</span><span class="op">)</span></span>
<span><span class="va">degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span><span class="va">scrna_orthoe</span>, group.by <span class="op">=</span> <span class="st">"JAK2_p.V617F_c.1849G.T_consensus"</span>, ident.1 <span class="op">=</span> <span class="st">"Alt"</span>, ident.2 <span class="op">=</span> <span class="st">"Ref"</span>, logfc.threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">degs</span>, gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">degs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">degs_up</span>   <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">degs</span>, <span class="va">avg_log2FC</span> <span class="op">&gt;</span>   <span class="fl">0.25</span> <span class="op">&amp;</span> <span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">degs_down</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">degs</span>, <span class="va">avg_log2FC</span> <span class="op">&lt;</span>  <span class="op">-</span><span class="fl">0.25</span> <span class="op">&amp;</span> <span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EnhancedVolcano/man/EnhancedVolcano.html" class="external-link">EnhancedVolcano</a></span><span class="op">(</span><span class="va">degs</span>, x <span class="op">=</span> <span class="st">"avg_log2FC"</span>, y <span class="op">=</span> <span class="st">"p_val_adj"</span>, lab <span class="op">=</span> <span class="va">degs</span><span class="op">$</span><span class="va">gene</span>, FCcutoff <span class="op">=</span> <span class="fl">0.25</span>, pCutoff <span class="op">=</span> <span class="fl">0.05</span>, title <span class="op">=</span> <span class="st">"OrthoE JAK2V617F Pos. vs. Neg."</span>,subtitle <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p><img src="MPN_SIGURD_files/figure-html/DEGs_OrthoE_JAK2V617F-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="combining-the-jak2v617f-data-with-the-mt-data-">Combining the JAK2V617F data with the MT data.<a class="anchor" aria-label="anchor" href="#combining-the-jak2v617f-data-with-the-mt-data-"></a>
</h2>
<p>We combine the JAK2V617F SIGURD object with the MT object. Cells and variants in both objects are combined.</p>
<p>To save time, we are only processing a single sample. All other samples are processed equally. Since the genotyping object is a list, it is easily processed using parallel computing.</p>
<p>First, we load and filter the mitochondrial variant information.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Loading the data. </span></span>
<span><span class="co"># genotyping_mt &lt;- LoadingMAEGATK_typewise(patient = "MPN1", samples_file = "/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/data/DesignMatrix_MT.csv", min_cells = 0, cells_include = colnames(scrna), type_use = "scRNAseq_MT", verbose = FALSE)</span></span>
<span><span class="co"># genotyping_mt &lt;- Filtering(genotyping_mt, min_variants_per_cell = 1, min_cells_per_variant = 2, fraction_threshold = 0.05, cells_include = colnames(scrna), verbose = FALSE)</span></span>
<span><span class="co"># Adding the JAK2V617F information.</span></span>
<span><span class="co"># genotyping_MPN1 &lt;- Filtering(genotyping, cells_include = grep("MPN1", colnames(genotyping), value = TRUE))</span></span>
<span><span class="co"># genotyping_mt &lt;- CombineSEobjects(genotyping_MPN1, genotyping_mt)</span></span>
<span></span>
<span><span class="co"># We load the Zenodo File. This is equivalent to the way we do it above. </span></span>
<span><span class="va">genotyping_mt_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_object.html">load_object</a></span><span class="op">(</span><span class="st">"/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/Paper_Figures/ZenodoFiles/scrna_genotyping.rds"</span><span class="op">)</span></span>
<span><span class="va">genotyping_mt</span> <span class="op">&lt;-</span> <span class="va">genotyping_mt_all</span><span class="op">[[</span><span class="st">"MPN1"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">genotyping_mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Filtering.html">Filtering</a></span><span class="op">(</span><span class="va">genotyping_mt</span>, min_variants_per_cell <span class="op">=</span> <span class="fl">1</span>, min_cells_per_variant <span class="op">=</span> <span class="fl">2</span>, fraction_threshold <span class="op">=</span> <span class="fl">0.05</span>, cells_include <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scrna</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="mpn-vois-and-diversity">MPN VOIs and Diversity<a class="anchor" aria-label="anchor" href="#mpn-vois-and-diversity"></a>
</h2>
<p>Now, we select the mtVOIs. We then determine the clonal lineages and plot the results on a heatmap.</p>
<p>We subset the cells to only include the erythroid cells.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Subset to erythroid cells.</span></span>
<span><span class="va">erythroid_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">scrna</span>, <span class="va">celltype_merged</span> <span class="op">==</span> <span class="st">"ErythroidCells"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">genotyping_mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Filtering.html">Filtering</a></span><span class="op">(</span><span class="va">genotyping_mt</span>, cells_include <span class="op">=</span> <span class="va">erythroid_cells</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "We remove all cells not in the allow list."</span></span>
<span><span class="co">## [1] "We remove all the variants that are always NoCall."</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Selecting the mtVOIS.</span></span>
<span><span class="va">mt_vois</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VariantSelection_Quantile.html">VariantSelection_Quantile</a></span><span class="op">(</span><span class="va">genotyping_mt</span>, min_coverage <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Determining the clonal lineages.</span></span>
<span><span class="va">genotyping_mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ClonalDefinition.html">ClonalDefinition</a></span><span class="op">(</span>se <span class="op">=</span> <span class="va">genotyping_mt</span>, variants_ls <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">mt_vois</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualisation-of-mtvois-and-clonal-lineages">Visualisation of mtVOIs and Clonal Lineages<a class="anchor" aria-label="anchor" href="#visualisation-of-mtvois-and-clonal-lineages"></a>
</h2>
<p>We now plot the variants of interest and the clonal lineages.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generating a heatmap.</span></span>
<span><span class="va">colors_lineage</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C1"</span> <span class="op">=</span> <span class="st">"#563e00"</span>, <span class="st">"C2"</span> <span class="op">=</span> <span class="st">"#dec234"</span>, <span class="st">"C3"</span> <span class="op">=</span> <span class="st">"#e23856"</span>, <span class="st">"C4"</span> <span class="op">=</span> <span class="st">"#7f0073"</span>, <span class="st">"C5"</span> <span class="op">=</span> <span class="st">"#006639"</span>, <span class="st">"C6"</span> <span class="op">=</span> <span class="st">"#b1db8e"</span>, <span class="st">"C7"</span> <span class="op">=</span> <span class="st">"#00348f"</span>, <span class="st">"Negative"</span> <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span>
<span><span class="va">ha</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/columnAnnotation.html" class="external-link">columnAnnotation</a></span><span class="op">(</span>Clones <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">genotyping_mt</span><span class="op">)</span><span class="op">[</span>,<span class="st">"Clones"</span><span class="op">]</span>,</span>
<span>                       col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Clones <span class="op">=</span> <span class="va">colors_lineage</span><span class="op">)</span>,</span>
<span>                       annotation_name_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">10</span>, fontface <span class="op">=</span> <span class="st">"plain"</span><span class="op">)</span>,</span>
<span>                       annotation_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Clones <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">1</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fraction</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">genotyping_mt</span><span class="op">)</span><span class="op">[[</span><span class="st">"fraction"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="va">mt_vois</span>, <span class="op">]</span></span>
<span><span class="va">hm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">fraction</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"VAF"</span>,</span>
<span>  row_title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>  row_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>  row_names_max_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">16</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  column_title <span class="op">=</span> <span class="st">"MPN1"</span>,</span>
<span>  column_title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>  column_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">fraction</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>, length.out <span class="op">=</span> <span class="fl">9</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#FCFCFC"</span>, <span class="st">"#FFEDB0"</span>, <span class="st">"#FFDF5F"</span>, <span class="st">"#FEC510"</span>, <span class="st">"#FA8E24"</span>, <span class="st">"#F14C2B"</span>, <span class="st">"#DA2828"</span>, <span class="st">"#BE2222"</span>, <span class="st">"#A31D1D"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  show_column_dend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_row_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cluster_columns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>border <span class="op">=</span> <span class="st">"#000000"</span>, grid_height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">10</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>  direction <span class="op">=</span> <span class="st">"horizontal"</span><span class="op">)</span>,</span>
<span>  bottom_annotation <span class="op">=</span> <span class="va">ha</span>,</span>
<span>  border <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use_raster <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_heatmap_legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">hm</span>, annotation_legend_side <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="MPN_SIGURD_files/figure-html/PlottingLineages_MPN1-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="calculating-clonal-diversity-">Calculating clonal diversity.<a class="anchor" aria-label="anchor" href="#calculating-clonal-diversity-"></a>
</h2>
<p>We calculate the clonal diversity for the sample MPN1 as an example.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clonal_diversity</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ClonalDiversity.html">ClonalDiversity</a></span><span class="op">(</span><span class="va">genotyping_mt</span>, grouping <span class="op">=</span> <span class="st">"Clones"</span>, diversity_measure <span class="op">=</span> <span class="st">"EffectiveSpecies"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Clonal Diversity in Effective Number of Species for MPN1: "</span>, <span class="va">clonal_diversity</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Clonal Diversity in Effective Number of Species for MPN1: 6.72545022391521"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="repeating-the-mt-analysis-for-all-samples-">Repeating the MT analysis for all samples.<a class="anchor" aria-label="anchor" href="#repeating-the-mt-analysis-for-all-samples-"></a>
</h2>
<p>We now repeat the MT analysis for all samples. This will take considereable time.</p>
<p>We speed the analsyis up by using parallel computing. Since the Zenodo file was already loaded, we save time here.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We get all patients present in the design matrix.</span></span>
<span><span class="va">design_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/data/DesignMatrix_MT.csv"</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span></span>
<span><span class="va">patients</span> <span class="op">&lt;-</span> <span class="va">design_matrix</span><span class="op">$</span><span class="va">patient</span></span>
<span><span class="co"># genotyping_mt_all &lt;- parallel::mclapply(patients, LoadingMAEGATK_typewise, samples_file = "/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/data/DesignMatrix_MT.csv", min_cells = 0, cells_include = colnames(scrna), type_use = "scRNAseq_MT", verbose = FALSE, mc.cores = length(patients))</span></span>
<span><span class="co"># names(genotyping_mt_all) &lt;- patients</span></span>
<span><span class="co"># We already loaded the data.</span></span>
<span><span class="va">genotyping_mt_all</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span><span class="va">genotyping_mt_all</span>, <span class="va">Filtering</span>, min_variants_per_cell <span class="op">=</span> <span class="fl">1</span>, min_cells_per_variant <span class="op">=</span> <span class="fl">2</span>, fraction_threshold <span class="op">=</span> <span class="fl">0.05</span>, cells_include <span class="op">=</span> <span class="va">erythroid_cells</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, mc.cores <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">patients</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="selecting-variants-of-interest-for-all-samples">Selecting variants of interest for all samples<a class="anchor" aria-label="anchor" href="#selecting-variants-of-interest-for-all-samples"></a>
</h2>
<p>We select variants of interest for all samples.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Selecting the mtVOIS.</span></span>
<span><span class="va">mt_vois_all</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span><span class="va">genotyping_mt_all</span>, <span class="va">VariantSelection_Quantile</span>, min_coverage <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, mc.cores <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">patients</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Determining the clonal lineages. This takes very long. We use parallel computing to process all samples simultaneously.</span></span>
<span><span class="va">genotyping_mt_all</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">mt_vois_all</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">sample_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mt_vois_all</span><span class="op">)</span><span class="op">[</span><span class="va">x</span><span class="op">]</span></span>
<span>  <span class="va">clones</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ClonalDefinition.html">ClonalDefinition</a></span><span class="op">(</span>se <span class="op">=</span> <span class="va">genotyping_mt_all</span><span class="op">[[</span><span class="va">sample_use</span><span class="op">]</span><span class="op">]</span>, <span class="va">mt_vois_all</span><span class="op">[</span><span class="va">sample_use</span><span class="op">]</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">clones</span><span class="op">)</span></span>
<span><span class="op">}</span>, mc.cores <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">mt_vois_all</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">genotyping_mt_all</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">patients</span></span>
<span><span class="fu"><a href="../reference/save_object.html">save_object</a></span><span class="op">(</span><span class="va">genotyping_mt_all</span>, <span class="st">"/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/Paper_Figures/ZenodoFiles/genotyping_mt_all.rds.lz4"</span>, <span class="st">"lz4"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculating-the-diverstiy-for-all-samples">Calculating the diverstiy for all samples<a class="anchor" aria-label="anchor" href="#calculating-the-diverstiy-for-all-samples"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We load the previous results.</span></span>
<span><span class="va">genotyping_mt_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_object.html">load_object</a></span><span class="op">(</span><span class="st">"/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/Paper_Figures/ZenodoFiles/genotyping_mt_all.rds.lz4"</span><span class="op">)</span></span>
<span><span class="va">clonal_diversity_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">genotyping_mt_all</span>, <span class="va">ClonalDiversity</span>, grouping <span class="op">=</span> <span class="st">"Clones"</span>, diversity_measure <span class="op">=</span> <span class="st">"EffectiveSpecies"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">clonal_diversity_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">genotyping_mt_all</span><span class="op">)</span>, EffectiveSpecies <span class="op">=</span> <span class="va">clonal_diversity_all</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">clonal_diversity_all</span>, format<span class="op">=</span><span class="st">"html"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Sample
</th>
<th style="text-align:right;">
EffectiveSpecies
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
MPN1
</td>
<td style="text-align:left;">
MPN1
</td>
<td style="text-align:right;">
6.72545
</td>
</tr>
<tr>
<td style="text-align:left;">
MPN2
</td>
<td style="text-align:left;">
MPN2
</td>
<td style="text-align:right;">
35.77652
</td>
</tr>
<tr>
<td style="text-align:left;">
MPN3
</td>
<td style="text-align:left;">
MPN3
</td>
<td style="text-align:right;">
24.17242
</td>
</tr>
<tr>
<td style="text-align:left;">
HC1
</td>
<td style="text-align:left;">
HC1
</td>
<td style="text-align:right;">
63.41134
</td>
</tr>
<tr>
<td style="text-align:left;">
HC2
</td>
<td style="text-align:left;">
HC2
</td>
<td style="text-align:right;">
873.95614
</td>
</tr>
<tr>
<td style="text-align:left;">
HC3
</td>
<td style="text-align:left;">
HC3
</td>
<td style="text-align:right;">
216.72850
</td>
</tr>
<tr>
<td style="text-align:left;">
MPN6
</td>
<td style="text-align:left;">
MPN6
</td>
<td style="text-align:right;">
10.98086
</td>
</tr>
<tr>
<td style="text-align:left;">
MPN4
</td>
<td style="text-align:left;">
MPN4
</td>
<td style="text-align:right;">
14.85520
</td>
</tr>
<tr>
<td style="text-align:left;">
MPN5
</td>
<td style="text-align:left;">
MPN5
</td>
<td style="text-align:right;">
42.10990
</td>
</tr>
</tbody>
</table>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Grasshoff, Ivan Costa.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
