<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="sigurd">
<title>Cell Mixture Analysis using Seq-Well Sequencing â€¢ sigurd</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Cell Mixture Analysis using Seq-Well Sequencing">
<meta property="og:description" content="sigurd">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sigurd</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.14</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/BoneMarrow_SIGURD.html">Analysis of Blastic Plasmacytoid Dendritic Cell Neoplasm (BPDCN)</a>
    <a class="dropdown-item" href="../articles/MPN_SIGURD.html">Analysis of MPN Data</a>
    <a class="dropdown-item" href="../articles/SW_CellMixture_SIGURD.html">Cell Mixture Analysis using Seq-Well Sequencing</a>
    <a class="dropdown-item" href="../articles/TenX_CellMixture_SIGURD.html">Cell Mixture Experiment using 10X Genomics Sequencing</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">

<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Cell Mixture Analysis using Seq-Well Sequencing</h1>
            
      
      
      <div class="d-none name"><code>SW_CellMixture_SIGURD.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="cell-mixture-analysis-using-sigurd">Cell Mixture Analysis using SIGURD<a class="anchor" aria-label="anchor" href="#cell-mixture-analysis-using-sigurd"></a>
</h2>
<p>In this data set, chronic myelogenous leukemia cells (K562) and brain tumor cells (BT142) were mixed and then sequenced using Seq-Well S3. The cells can be separated using the expression of marker genes.</p>
<p>The original script then identifies 6 variants that can distinguish between the two cell types.</p>
<p>In this reproduction, we show that the analysis can be faithfully reproduced using a very streamlined set of functions. This greatly simplifies the analysis. Using a set of dedicated functions, the analysis can now also be easily adopted to new data sets.</p>
</div>
<div class="section level2">
<h2 id="loading-necessary-packages-">Loading necessary packages.<a class="anchor" aria-label="anchor" href="#loading-necessary-packages-"></a>
</h2>
<p>We load the data from MAESTER.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Libraries for SIGURD."</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Libraries for SIGURD."</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://costalab.github.io/sigurd/">sigurd</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="we-load-the-seurat-object-for-the-scrna-seq-data-">We load the Seurat object for the scRNA-seq data.<a class="anchor" aria-label="anchor" href="#we-load-the-seurat-object-for-the-scrna-seq-data-"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/data/SW_CellLineMix_Seurat_Keep_Renamed.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-the-data-using-sigurd-">Loading the data using SIGURD.<a class="anchor" aria-label="anchor" href="#loading-the-data-using-sigurd-"></a>
</h2>
<p>SIGURD loads the data and automatically reformats the data. This code uses dedicated functions to load the and process the data, which streamlines the analysis quite substantially.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The design matrix contains information for the genotyping data.</span></span>
<span><span class="va">genotyping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadingMAEGATK_typewise.html">LoadingMAEGATK_typewise</a></span><span class="op">(</span>patient <span class="op">=</span> <span class="st">"SW"</span>, samples_file <span class="op">=</span> <span class="st">"/data/MPN/exp/scRNA/MPN_mutations/SIGURD_paper/sigurd/data/MAESTER_Reproduction.csv"</span>, min_cells <span class="op">=</span> <span class="fl">0</span>, type_use <span class="op">=</span> <span class="st">"Amplicon_MT"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Removing cells that are not in the Seurat object.</span></span>
<span><span class="va">genotyping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Filtering.html">Filtering</a></span><span class="op">(</span><span class="va">genotyping</span>, cells_include <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "We remove all cells not in the allow list."</span></span>
<span><span class="co">## [1] "We remove all the variants that are always NoCall."</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Adding meta data do the SIGURD object.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">genotyping</span><span class="op">)</span><span class="op">$</span><span class="va">CellType</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">$</span><span class="va">CellType</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="selecting-the-variants-of-interest">Selecting the Variants of Interest<a class="anchor" aria-label="anchor" href="#selecting-the-variants-of-interest"></a>
</h2>
<p>Now, we select the variants of interest.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Selecting informative variants. </span></span>
<span><span class="va">voi.ch.sigurd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VariantSelection_Quantile.html">VariantSelection_Quantile</a></span><span class="op">(</span><span class="va">genotyping</span>, min_coverage <span class="op">=</span> <span class="fl">20</span>, min_quality <span class="op">=</span> <span class="fl">30</span>, quantiles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.9</span><span class="op">)</span>, thresholds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.9</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">voi.ch.sigurd</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "chrM_709_G_A"  "chrM_1888_G_A" "chrM_1420_T_C" "chrM_2141_T_C"</span></span>
<span><span class="co">## [5] "chrM_9117_T_C" "chrM_7990_C_T"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="visualisation-using-sigurd">Visualisation using SIGURD<a class="anchor" aria-label="anchor" href="#visualisation-using-sigurd"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Determining if a cell is supporting one of the two cell types.</span></span>
<span><span class="va">cell_support</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CallSupport.html">CallSupport</a></span><span class="op">(</span>SE <span class="op">=</span> <span class="va">genotyping</span>, VOI_group1 <span class="op">=</span> <span class="va">voi.ch.sigurd</span><span class="op">[</span><span class="va">voi.ch.sigurd</span> <span class="op">!=</span> <span class="st">"chrM_7990_C_T"</span><span class="op">]</span>, VOI_group2 <span class="op">=</span> <span class="st">"chrM_7990_C_T"</span>, group1_name <span class="op">=</span> <span class="st">"K562"</span>, group2_name <span class="op">=</span> <span class="st">"BT142"</span>, min_mutated_reads <span class="op">=</span> <span class="fl">3</span>, min_reads <span class="op">=</span> <span class="fl">30</span>, group_factor <span class="op">=</span> <span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, return_nonsupport <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">genotyping</span><span class="op">)</span><span class="op">$</span><span class="va">CellType_MT</span> <span class="op">&lt;-</span> <span class="va">cell_support</span><span class="op">$</span><span class="va">Support</span></span>
<span><span class="va">cell_support</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">cell_support</span>, <span class="va">Support</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"K562"</span>, <span class="st">"BT142"</span><span class="op">)</span><span class="op">)</span> <span class="co"># Selecting only cells with sufficient support for a set of variants.</span></span>
<span></span>
<span><span class="co"># Plotting the results on a heatmap.</span></span>
<span><span class="fu"><a href="../reference/HeatmapVoi.html">HeatmapVoi</a></span><span class="op">(</span>SE <span class="op">=</span> <span class="va">genotyping</span><span class="op">[</span>,<span class="va">cell_support</span><span class="op">$</span><span class="va">Cell</span><span class="op">]</span>, voi <span class="op">=</span> <span class="va">voi.ch.sigurd</span>, annotation_trait <span class="op">=</span> <span class="st">"CellType"</span>, minimum_coverage <span class="op">=</span> <span class="fl">3</span>, remove_empty_cells <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SW_CellMixture_SIGURD_files/figure-html/unnamed-chunk-4-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="selecting-variants-with-comparatively-high-vaf-between-groups">Selecting variants with comparatively high VAF between groups<a class="anchor" aria-label="anchor" href="#selecting-variants-with-comparatively-high-vaf-between-groups"></a>
</h2>
<p>We now select variants with a higher VAF in the K562 cells than in the BT142 cells.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">voi.ch.lineages</span> <span class="op">&lt;-</span> <span class="fu">sigurd</span><span class="fu">::</span><span class="fu"><a href="../reference/VariantSelection_Group.html">VariantSelection_Group</a></span><span class="op">(</span>SE <span class="op">=</span> <span class="va">genotyping</span>, min_coverage <span class="op">=</span> <span class="fl">20</span>, quantiles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.99</span><span class="op">)</span>, thresholds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.02</span><span class="op">)</span>, min_quality <span class="op">=</span> <span class="fl">30</span>,</span>
<span>                                                  group_of_interest <span class="op">=</span> <span class="st">"CellType_MT"</span>, group1 <span class="op">=</span> <span class="st">"K562"</span>, group2 <span class="op">=</span> <span class="st">"BT142"</span>, group_factor <span class="op">=</span> <span class="fl">5</span>, remove_nocall <span class="op">=</span> <span class="cn">FALSE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">voi.ch.lineages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"chrM_9117_T_C"</span>, <span class="st">"chrM_2141_T_C"</span>, <span class="va">voi.ch.lineages</span><span class="op">)</span> <span class="co"># Replace chrM_9117_T_C, which comes up because it has no coverage in some cells, with another homoplasmic variant (chrM_2141_T_C), that has better coverage)</span></span>
<span><span class="va">voi.ch.lineages</span> <span class="op">&lt;-</span> <span class="va">voi.ch.lineages</span><span class="op">[</span><span class="op">!</span><span class="va">voi.ch.lineages</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chrM_5378_A_G"</span>, <span class="st">"chrM_6384_G_A"</span><span class="op">)</span><span class="op">]</span> <span class="co"># These variants are not detected in the bulk data and were removed by hand.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">voi.ch.lineages</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "chrM_1197_G_A" "chrM_1227_G_A" "chrM_1327_G_A" "chrM_1982_G_A"</span></span>
<span><span class="co">##  [5] "chrM_2008_G_A" "chrM_2147_G_A" "chrM_2170_G_A" "chrM_2478_G_A"</span></span>
<span><span class="co">##  [9] "chrM_2571_G_A" "chrM_2573_G_A" "chrM_2810_G_A" "chrM_2943_G_A"</span></span>
<span><span class="co">## [13] "chrM_5274_G_A" "chrM_8155_G_A" "chrM_8213_G_A" "chrM_8278_C_A"</span></span>
<span><span class="co">## [17] "chrM_9778_G_A" "chrM_2141_T_C" "chrM_1310_C_T" "chrM_1352_C_T"</span></span>
<span><span class="co">## [21] "chrM_1533_C_T" "chrM_9757_C_T"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="visualisation-using-sigurd-1">Visualisation using SIGURD<a class="anchor" aria-label="anchor" href="#visualisation-using-sigurd-1"></a>
</h2>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plotting the results on a heatmap.</span></span>
<span><span class="va">cell_support</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">cell_support</span>, <span class="va">Support</span> <span class="op">==</span> <span class="st">"K562"</span><span class="op">)</span></span>
<span><span class="va">genotyping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Filtering.html">Filtering</a></span><span class="op">(</span><span class="va">genotyping</span>, cells_include <span class="op">=</span> <span class="va">cell_support</span><span class="op">$</span><span class="va">Cell</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/HeatmapVoi.html">HeatmapVoi</a></span><span class="op">(</span>SE <span class="op">=</span> <span class="va">genotyping</span>, voi <span class="op">=</span> <span class="va">voi.ch.lineages</span>, minimum_coverage <span class="op">=</span> <span class="fl">3</span>, sort_cells <span class="op">=</span> <span class="cn">TRUE</span>, cluster_variants <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SW_CellMixture_SIGURD_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Grasshoff, Ivan Costa.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
