<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="We filter a SummarizedExperiment object to exclude variants and cells."><title>Filtering the loaded genotyping data. — Filtering • sigurd</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Filtering the loaded genotyping data. — Filtering"><meta property="og:description" content="We filter a SummarizedExperiment object to exclude variants and cells."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sigurd</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.14</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/BoneMarrow_SIGURD.html">Analysis of Blastic Plasmacytoid Dendritic Cell Neoplasm (BPDCN)</a>
    <a class="dropdown-item" href="../articles/MPN_SIGURD.html">Analysis of MPN Data</a>
    <a class="dropdown-item" href="../articles/SW_CellMixture_SIGURD.html">Cell Mixture Analysis using Seq-Well Sequencing</a>
    <a class="dropdown-item" href="../articles/TenX_CellMixture_SIGURD.html">Cell Mixture Experiment using 10X Genomics Sequencing</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Filtering the loaded genotyping data.</h1>
      
      <div class="d-none name"><code>Filtering.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>We filter a SummarizedExperiment object to exclude variants and cells.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">Filtering</span><span class="op">(</span></span>
<span>  <span class="va">SE</span>,</span>
<span>  cells_include <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cells_exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fraction_threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  keep_refs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  alts_threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_cells_per_variant <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_variants_per_cell <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reject_value <span class="op">=</span> <span class="st">"NoCall"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>SE</dt>
<dd><p>SummarizedExperiment object.</p></dd>


<dt>cells_include</dt>
<dd><p>A vector of cell barcodes. Only these cells will be retained.</p></dd>


<dt>cells_exclude</dt>
<dd><p>A vector of cell barcodes. These cells will be removed from the output.</p></dd>


<dt>fraction_threshold</dt>
<dd><p>Variants with an VAF below this threshold are set to 0. Numeric. Default = NULL.</p></dd>


<dt>keep_refs</dt>
<dd><p>If a cell is below the fraction_threshold, should the number of reference reads be kept? If FALSE, the reference reads will be removed. Default = TRUE</p></dd>


<dt>alts_threshold</dt>
<dd><p>Variants with a number of alt reads less than this threshold are set to 0. Numeric. Default = NULL.</p></dd>


<dt>min_cells_per_variant</dt>
<dd><p>In how many cells should a variant be present to be included? Numeric. Default = 2.</p></dd>


<dt>min_variants_per_cell</dt>
<dd><p>How many variants should be covered in a cell have to be included? Default = 1.</p></dd>


<dt>reject_value</dt>
<dd><p>Should cells that fall below a threshold (fraction_threshold or alts_threshold) be treated as Reference or NoCall? Default = NoCall.</p></dd>


<dt>verbose</dt>
<dd><p>Should the function be verbose? Default = TRUE</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>We do this for one sample at a time.
We want to remove:</p><ol><li><p>all cells not in an allow list,</p></li>
<li><p>all cells in an exclusion list,</p></li>
<li><p>all cells that do not have at least one variant with &gt;1 (Reference),</p></li>
<li><p>all variants that are for alternative transcripts,</p></li>
<li><p>all variants that are always NoCall,</p></li>
<li><p>set variants with a VAF below a threshold to NoCall or Reference.</p></li>
</ol></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="co"># Removing all variants that are not detected in at least 2 cells.</span></span></span>
<span class="r-in"><span>  <span class="co"># Before we remove the variants, we set fraction value of variants below 0.05 to 0.</span></span></span>
<span class="r-in"><span>  <span class="va">se_geno</span> <span class="op">&lt;-</span> <span class="fu">Filtering</span><span class="op">(</span><span class="va">se_geno</span>, min_cells_per_variant <span class="op">=</span> <span class="fl">2</span>, fraction_threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Martin Grasshoff, Ivan Costa.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

